R/filter-nm-data.R

```diff
@@ -0,0 +1,193 @@
+
+  # Handle single `=` only when it's not part of `==`, `!=`, `<=`, `>=`
+  expr <- gsub("(?<=[^=!<>])=(?=[^=!<>])", "==", expr, perl = TRUE)
+  return(expr)
+}
+
+#' Function to invert ignore expressions
+#' @noRd
+#' @keywords internal
+invert_operator <- function(expr) {
+  expr <- dplyr::case_when(
+    grepl("==", expr) ~ gsub("==", "!=", expr),
+    grepl("!=", expr) ~ gsub("!=", "==", expr),
+    grepl("<=", expr) ~ gsub("<=", ">=", expr),
+    grepl(">=", expr) ~ gsub(">=", "<=", expr),
+    grepl("<", expr) ~ gsub("<", ">", expr),
+    grepl(">", expr) ~ gsub(">", "<", expr),
+    TRUE ~ expr
+  )
+  return(expr)
+}
+
+#' Translate `IGNORE` and `ACCEPT` filter expressions into [dplyr::filter()]
+#'  expressions.
+#'
+#' @param nm_expr a `NONMEM` filter expression. e.g., `'ID.EQ.2, BLQ=1'`.
+#' @param type Either `'ignore'` or `'accept'`. Denotes which type of `NONMEM`
+#'  filtering the expression corresponds to.
+#' @param data_cols column names associated with the input data. Used for
+#'  `'ignore'` expressions.
+#'
+#' @examples
+#' \dontrun{
+#' test_exprs <- "SEX==1, ID.EQ.2, WT/=70, AGE.NE.30, A=1, WT.GT.40"
+#'
+#' translate_nm_expr(test_exprs, type = 'ignore')
+#' #> [1] "SEX!=1 & ID!=2 & WT==70 & AGE==30 & A!=1 & WT<40"
+#'
+#' translate_nm_expr(test_exprs, type = 'accept')
+#' #> [1] "SEX==1 & ID==2 & WT!=70 & AGE!=30 & A==1 & WT>40"
+#'
+#'
+#' # Use of `@`, `#`, or form `IGNORE=C2` require `data_cols` to be specified,
+#' # though only the first column is used
+#' data_cols <- c("C", "ID", "TIME", "EVID", "DV", "BLQ")
+#'
+#' translate_nm_expr("#", data_cols = data_cols)
+#' #> [1] "!grepl('^#', C)"
+#'
+#' translate_nm_expr("c2", data_cols = data_cols)
+#' #> [1] "C!='c2'"
+#'
+#' translate_nm_expr("@", data_cols = data_cols)
+#' #> [1] "!grepl('^[A-Za-z@]', C)"
+#' }
+#' @keywords internal
+translate_nm_expr <- function(
+    nm_expr,
+    type = c("ignore", "accept"),
+    data_cols = NULL
+){
+  type <- match.arg(type)
+  checkmate::assert_character(data_cols, min.len = 1, null.ok = TRUE)
+
+  # Translate NM operators and separate any (list) type expressions
+  exprs <- translate_nm_operator(nm_expr)
+  exprs <- stringr::str_split(exprs, ",")[[1]] %>% stringr::str_trim()
+
+  r_exprs <- purrr::map_chr(exprs, function(expr){
+    if(type == "ignore"){
+      # `IGNORE=#`, `IGNORE=@`, `IGNORE=c1`, `IGNORE=(list)`
+      if(expr == "#"){
+        # IGNORE=# is the default.  That is, in the absence of IGNORE option, any
+        # record whose first character is '#' is treated as a comment record.
+        paste0("!grepl('^#', ", data_cols[1], ")")
+      }else if(expr == "@"){
+        # IGNORE=@ signifies that any data record having an alphabetic character
+        # or `@` as its first non-blank character in column one should be ignored.
+        #  - This permits a table file having header lines to be used as an
+        #    NM-TRAN data set.
+        paste0("!grepl('^[A-Za-z@]', ", data_cols[1], ")")
```

Comment:
I thought about it since you had that, but didnt see anything in the `NONMEM` docs that would indicate that behavior. It specifically says:
 > IGNORE=@  signifies  that  any  data  record having an alphabetic character or @ as its first non-blank character (not just in column  1) should be ignored.

Reply:
I was suggesting it because of what you shared in the docs ... first _non-blank_ character.  If `data_cols[1]` had a row that was ` C`, it should be ignored, right?

Reply:
Yeah that sounds right; just made the adjustment